rosetta/transactor/parser_test.go

```diff
@@ -0,0 +1,613 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package transactor_test
+
+import (
+	"crypto/rand"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/cadence"
+	cjson "github.com/onflow/cadence/encoding/json"
+	sdk "github.com/onflow/flow-go-sdk"
+	sdkcrypto "github.com/onflow/flow-go-sdk/crypto"
+	"github.com/onflow/flow-go/crypto"
+	chash "github.com/onflow/flow-go/crypto/hash"
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/rosetta/failure"
+	"github.com/optakt/flow-dps/rosetta/identifier"
+	"github.com/optakt/flow-dps/rosetta/transactor"
+	"github.com/optakt/flow-dps/testing/mocks"
+)
+
+func TestTransactionParser_BlockID(t *testing.T) {
+	header := mocks.GenericHeader
+	blockID := header.ID()
+	index := uint64(84)
+
+	t.Run("nominal case", func(t *testing.T) {
+		t.Parallel()
+
+		tx := &sdk.Transaction{
+			ReferenceBlockID: sdk.HashToID(blockID[:]),
+		}
+
+		validator := mocks.BaselineValidator(t)
+		validator.BlockFunc = func(rosBlockID identifier.Block) (uint64, flow.Identifier, error) {
+			assert.Equal(t, tx.ReferenceBlockID.String(), rosBlockID.Hash)
+
+			return index, blockID, nil
+		}
+
+		p := transactor.BaselineTransactionParser(
+			t,
+			transactor.ParseTransaction(tx),
+			transactor.ParseValidator(validator),
```

Comment:
Definitely seems to me that `WithTransaction` and `WithValidator` are the better choices here.

Reply:
As mentioned in the PR body, unfortunately the problem is that transactor has two structures that can be built with `BaselineXXX`, and they both have the same dependencies to inject. Thus, the transactor has `WithXXX` for all of its dependencies, while I couldn't do the same with the `Parser`. I'm open to suggestions, I mean technically we could even have options that work for both since they have the same attributes, make the option functions take an `interface{}` and cast it into the proper type to set their attributes.

Do you think that would be cleaner?